tools/ci_build/rewriter.py

```diff
@@ -0,0 +1,221 @@
+'''comment out unused ops in cpu_execution_provider.cc by a set of models'''
+
+import os
+import shutil
+import onnx
+from onnx import AttributeProto as AP
+
+#pylint: disable=no-member,too-many-locals,too-many-statements
+
+def extract_ops_from_file(file_path, referred_ops):
+    '''extract ops from csv file with format:
+        op_type,domain,opset'''
+
+    if not os.path.isfile(file_path):
+        return referred_ops
+
+    with open(file_path, 'r') as csv_to_read:
+
+        for line in csv_to_read.readlines():
+            op_type, domain, raw_opset = line.strip().split(',')
+            opset = int(raw_opset)
+
+            if op_type not in referred_ops:
+                referred_ops[op_type] = {domain:[opset]}
+
+            elif domain not in referred_ops[op_type]:
+                referred_ops[op_type][domain] = [opset]
+
+            elif opset not in referred_ops[op_type][domain]:
+                referred_ops[op_type][domain].append(opset)
+
+    return referred_ops #end of extract_ops_from_file(...)
+
+
+def extract_ops_from_model(model_path, referred_ops):
+    '''extract ops from models under model_path and return a diction'''
+
+    if not os.path.isdir(model_path):
+        return referred_ops
+
+    def map_domain(domain):
+        if domain == 'ai.onnx.ml':
+            return 'kMLDomain'
+        return 'kOnnxDomain'
+
+    def extract_ops_from_graph(graph, opsets, operators):
+        '''extract ops from graph and all subgraphs'''
+
+        for operator in graph.node:
+
+            if operator.op_type not in operators:
+                operators[operator.op_type] = {}
+
+            for opset in opsets:
+                mapped_domain = map_domain(opset.domain)
+                if mapped_domain not in operators[operator.op_type]:
+                    operators[operator.op_type][mapped_domain] = []
+                if opset.version not in operators[operator.op_type][mapped_domain]:
+                    operators[operator.op_type][mapped_domain].append(opset.version)
+
+            for attr in operator.attribute:
+
+                if attr.type == AP.GRAPH: #process subgraph
+                    extract_ops_from_graph(attr.g, opsets, operators)
+
+                elif attr.type == AP.GRAPHS: #process all subgraphs
+                    for subgraph in attr.graphs:
+                        extract_ops_from_graph(subgraph, opsets, operators)
+
+    for root, _, files in os.walk(model_path):
+        for file in files:
+            if file.endswith('.onnx'):
+                model_path = os.path.join(root, file)
+                model = onnx.load(model_path)
+                extract_ops_from_graph(model.graph, model.opset_import, referred_ops)
+
+    return referred_ops #end of extract_ops_from_model(...)
+
+
+def rewrite_cpu_provider(model_path, file_path, ep_path):
```

Comment:
This is no longer limited to cpu only, right?

Reply:
Exactly, renaming the function...